File: web/lib/computeGrades.ts
/**
 * TypeScript port of pipeline/compute_grades.py's grading formula, used
 * only by the "Recompute grades now" button (app/api/recompute-grades).
 * The full pipeline run still does this in Python -- this is a deliberate
 * duplication so the button doesn't need a mixed Node+Python deployment.
 *
 * KEEP THIS IN SYNC WITH pipeline/compute_grades.py -- if the formula,
 * bands, or weighting ever change there, mirror the change here too.
 *
 * See compute_grades.py's docstring for the full plain-language writeup of
 * how the formula works. Short version: every raw stat is min-max scaled
 * to 0-100 across whichever teams are passed in (best=100, worst=0), then
 * blended into Pass Block (weighted 17/34/34/15: sack rate / pressure
 * rate / ESPN PBWR / pass Strength of Schedule) / Run Block (28.3/28.3/
 * 28.3/15: stuff rate / YBC per att / ESPN RBWR / run Strength of
 * Schedule) scores -- each of which then gets a SECOND min-max stretch
 * across all teams (otherwise averaging multiple components means almost
 * nobody ever hits a true 0 or 100, so A+/F sit empty). Overall is then a
 * weighted 46/46/8 blend of Pass Block / Run Block / penalty score (O-line
 * penalty rate, lower is better) -- NOT a plain average of Pass/Run --
 * getting its own final stretch, before every score maps to a letter on an
 * equal-width 13-band scale.
 */

export const GRADE_FORMULA_VERSION = "v5";

// Pass/Run Block component weights -- pressure rate and ESPN's win rate
// count double a raw sack (sacks are partly a QB-behavior/scheme stat, not
// purely an O-line one). These are relative proportions, not required to
// sum to 1 -- see weightedAverage.
const SACK_RATE_WEIGHT = 0.17;
const PRESSURE_RATE_WEIGHT = 0.34;
const ESPN_PBWR_WEIGHT = 0.34;
const PASS_SOS_WEIGHT = 0.15;

const RUN_BLOCK_COMPONENT_WEIGHT = 0.283; // stuff rate / YBC per att / ESPN RBWR, each
const RUN_SOS_WEIGHT = 0.15;

// Overall score weights -- small and deliberately so; penalty rate is a
// nudge, not a co-equal grading dimension (see the module docstring).
const OVERALL_PASS_WEIGHT = 0.46;
const OVERALL_RUN_WEIGHT = 0.46;
const OVERALL_PENALTY_WEIGHT = 0.08;

const GRADE_BANDS: [number, string][] = [
  [92.3, "A+"], [84.6, "A"], [76.9, "A-"],
  [69.2, "B+"], [61.5, "B"], [53.8, "B-"],
  [46.2, "C+"], [38.5, "C"], [30.8, "C-"],
  [23.1, "D+"], [15.4, "D"], [7.7, "D-"],
  [0, "F"],
];

export function scoreToLetter(score: number | null): string | null {
  if (score === null) return null;
  for (const [threshold, letter] of GRADE_BANDS) {
    if (score >= threshold) return letter;
  }
  return "F";
}

/** Min-max scales a column of values (one per team) to 0-100. Ties or an
 * all-null column give everyone a neutral 50 rather than dividing by 0. */
function normalize(values: (number | null)[], higherIsBetter: boolean): (number | null)[] {
  const present = values.filter((v): v is number => v !== null);
  if (present.length === 0) return values.map(() => null);

  const lo = Math.min(...present);
  const hi = Math.max(...present);
  if (lo === hi) return values.map((v) => (v === null ? null : 50));

  return values.map((v) => {
    if (v === null) return null;
    const scaled = ((v - lo) / (hi - lo)) * 100;
    return higherIsBetter ? scaled : 100 - scaled;
  });
}

/** Weighted average of whichever components are non-null for each team. A
 * missing component (e.g. ESPN data not entered yet) drops its weight out
 * of the denominator entirely, so the remaining components' *relative*
 * weights are preserved rather than the gap being silently averaged away
 * unweighted. Pass a weight of 1 for every component to get a plain
 * average. */
function weightedAverage(components: [(number | null)[], number][]): (number | null)[] {
  const count = components[0]?.[0].length ?? 0;
  const result: (number | null)[] = [];
  for (let i = 0; i < count; i++) {
    let weightedSum = 0;
    let weightSum = 0;
    for (const [values, weight] of components) {
      const v = values[i];
      if (v !== null) {
        weightedSum += v * weight;
        weightSum += weight;
      }
    }
    result.push(weightSum > 0 ? weightedSum / weightSum : null);
  }
  return result;
}

export type TeamRawStats = {
  team_abbr: string;
  sacks_allowed: number;
  dropbacks: number;
  pressure_rate_allowed: number | null;
  stuff_rate: number | null;
  yards_before_contact_per_att: number | null;
  penalty_rate: number | null;
};

export type EspnTeamRate = {
  team_abbr: string;
  pass_block_win_rate: number | null;
  run_block_win_rate: number | null;
};

export type TeamScheduleStrength = {
  team_abbr: string;
  pass_sos_score: number | null;
  run_sos_score: number | null;
};

export type GradedTeam = {
  team_abbr: string;
  pass_block_score: number | null;
  pass_block_grade: string | null;
  run_block_score: number | null;
  run_block_grade: string | null;
  overall_score: number | null;
  overall_grade: string | null;
  grade_formula_version: string;
};

/** Takes every team's latest raw stats for a season (one row per team) plus
 * whatever ESPN rates and Strength of Schedule have been computed, and
 * returns the three grades per team -- comparing each team against all the
 * others passed in. */
export function computeGrades(
  rawStats: TeamRawStats[],
  espnRates: EspnTeamRate[],
  scheduleStrength: TeamScheduleStrength[]
): GradedTeam[] {
  const espnByTeam = new Map(espnRates.map((r) => [r.team_abbr, r]));
  const sosByTeam = new Map(scheduleStrength.map((r) => [r.team_abbr, r]));

  const sackRates = rawStats.map((t) => (t.dropbacks > 0 ? t.sacks_allowed / t.dropbacks : null));
  const pressureRates = rawStats.map((t) => t.pressure_rate_allowed);
  const stuffRates = rawStats.map((t) => t.stuff_rate);
  const ybcPerAtt = rawStats.map((t) => t.yards_before_contact_per_att);
  const pbwr = rawStats.map((t) => espnByTeam.get(t.team_abbr)?.pass_block_win_rate ?? null);
  const rbwr = rawStats.map((t) => espnByTeam.get(t.team_abbr)?.run_block_win_rate ?? null);
  const passSos = rawStats.map((t) => sosByTeam.get(t.team_abbr)?.pass_sos_score ?? null);
  const runSos = rawStats.map((t) => sosByTeam.get(t.team_abbr)?.run_sos_score ?? null);

  const normalizedPbwr = pbwr.every((v) => v === null) ? rawStats.map(() => null) : normalize(pbwr, true);
  const normalizedRbwr = rbwr.every((v) => v === null) ? rawStats.map(() => null) : normalize(rbwr, true);
  const normalizedPassSos = passSos.every((v) => v === null) ? rawStats.map(() => null) : normalize(passSos, true);
  const normalizedRunSos = runSos.every((v) => v === null) ? rawStats.map(() => null) : normalize(runSos, true);

  const rawPassScores = weightedAverage([
    [normalize(sackRates, false), SACK_RATE_WEIGHT],
    [normalize(pressureRates, false), PRESSURE_RATE_WEIGHT],
    [normalizedPbwr, ESPN_PBWR_WEIGHT],
    [normalizedPassSos, PASS_SOS_WEIGHT],
  ]);
  const rawRunScores = weightedAverage([
    [normalize(stuffRates, false), RUN_BLOCK_COMPONENT_WEIGHT],
    [normalize(ybcPerAtt, true), RUN_BLOCK_COMPONENT_WEIGHT],
    [normalizedRbwr, RUN_BLOCK_COMPONENT_WEIGHT],
    [normalizedRunSos, RUN_SOS_WEIGHT],
  ]);

  // Stretch each blended score to span the full 0-100 range across all
  // teams passed in -- see the module docstring for why (otherwise
  // almost nobody ever hits a true A+ or F).
  const passScores = normalize(rawPassScores, true);
  const runScores = normalize(rawRunScores, true);

  const penaltyRates = rawStats.map((t) => t.penalty_rate);
  const normalizedPenalty = penaltyRates.every((v) => v === null) ? rawStats.map(() => null) : normalize(penaltyRates, false);

  const rawOverallScores = weightedAverage([
    [passScores, OVERALL_PASS_WEIGHT],
    [runScores, OVERALL_RUN_WEIGHT],
    [normalizedPenalty, OVERALL_PENALTY_WEIGHT],
  ]);
  const overallScores = normalize(rawOverallScores, true);

  return rawStats.map((team, i) => {
    const pass_block_score = passScores[i];
    const run_block_score = runScores[i];
    const overall_score = overallScores[i];

    return {
      team_abbr: team.team_abbr,
      pass_block_score,
      pass_block_grade: scoreToLetter(pass_block_score),
      run_block_score,
      run_block_grade: scoreToLetter(run_block_score),
      overall_score,
      overall_grade: scoreToLetter(overall_score),
      grade_formula_version: GRADE_FORMULA_VERSION,
    };
  });
}

File: pipeline/compute_grades.py
"""Turns raw O-line stats into the three letter grades shown on each team
page: Pass Block, Run Block, and Overall. This file has NO network calls --
it's a pure function of the DataFrames you pass it, which is what makes it
easy to unit-test (see tests/test_compute_grades.py) and safe to reason
about without touching a real database.

--- How the grading works (v5) ---

Every raw stat gets converted into a 0-100 "component score" by comparing
each team against every OTHER team that same week, using min-max scaling:
the best team in the league that week scores 100 on that stat, the worst
scores 0, and everyone else falls in between. This is a CURVED/RELATIVE
grade, not an absolute one -- a team's grade can shift slightly week to
week even if its own raw numbers don't change, because the league-wide
best/worst can move. That's a known, intentional tradeoff (see the
project plan's shortcuts list).

After blending components into Pass Block / Run Block / Overall (below),
each of those three scores gets a second min-max "stretch" across the
week's 32 teams before being mapped to a letter. Without this, averaging
2-3 already-normalized components mathematically compresses everyone
toward the middle -- almost no team is #1 (or dead last) on *every*
component simultaneously, so almost nobody ever actually hits a true 0 or
100, and the A+/F bands sit empty every week. Stretching guarantees the
week's actual best team hits exactly 100 (A+) and the actual worst hits
exactly 0 (F), same as any individual raw stat already does.

Pass Block score = WEIGHTED average of whichever of these are available
(weights below; a missing component's weight is dropped and the rest
renormalized, not just averaged unweighted -- see _weighted_average):
  - sack rate (sacks allowed / dropback), weight 0.17        -- lower is better
  - pressure rate allowed, weight 0.34                       -- lower is better
  - ESPN Pass Block Win Rate (manually entered), weight 0.34 -- higher is better
  - pass Strength of Schedule, weight 0.15                   -- higher is better
Pressure rate and ESPN's win rate count double a raw sack -- sacks are
partly a QB-behavior/scheme stat (how long they hold the ball, whether
they take a bigger loss trying to escape) rather than purely an O-line
one, so pressure rate is the more stable signal of protection quality.
Sacks still carry real weight since they're a real, costly outcome.

Run Block score = weighted average of whichever of these are available
(equal thirds among the automated/ESPN components, same as before, with
Strength of Schedule folded in at the same 0.15 as Pass Block):
  - stuff rate (rush attempts stopped at/behind the line), weight 0.283 -- lower is better
  - yards before contact per attempt, weight 0.283                     -- higher is better
  - ESPN Run Block Win Rate (manually entered), weight 0.283           -- higher is better
  - run Strength of Schedule, weight 0.15                              -- higher is better

If ESPN's number hasn't been entered yet for a team/season, that component
is just left out of the average -- the score still appears (from the
remaining components), and improves in accuracy once ESPN data is added.
Same for Strength of Schedule if a team's schedule_strength row is missing
(e.g. very early in a season before compute_schedule_strength.py has been
run) -- see pipeline/steps/pull_schedule_strength.py for how it's computed.

Overall score = weighted average of Pass Block score (46%), Run Block
score (46%), and a penalty score (8%) -- NOT a plain average of Pass/Run
like earlier versions. The penalty score is O-line penalty rate
(Offensive Holding + False Start per offensive play, see
pull_team_ol_stats.py) normalized the same min-max way as everything
else, lower is better. Deliberately small and deliberately invisible as
its own line on the team page -- it's meant to nudge Overall slightly for
teams that are chronically undisciplined, not to be a 4th headline grade.
Surfaced instead as a "Nth-worst penalty rate" note in the Overall grade
badge's hover tooltip (see web/lib/getTeamPageData.ts /
web/components/GradeBadge.tsx). Pass Block and Run Block scores are
completely unaffected by penalty rate.

All three scores map to letters using the same academic-style scale.
Bump GRADE_FORMULA_VERSION any time the weights or bands below change, so
old and new grades in the database are distinguishable.

KEEP THIS IN SYNC WITH web/lib/computeGrades.ts -- that's a TypeScript
port of this exact formula, used by the "Recompute grades now" button so
grades can refresh without waiting for a full pipeline run. If you change
the weights, bands, or components here, mirror the change there too.
"""

import pandas as pd

GRADE_FORMULA_VERSION = "v5"

# Pass/Run Block component weights -- see the module docstring for the
# reasoning. These are relative proportions, not required to sum to 1:
# the denominator in _weighted_average is always the sum of whichever
# weights are actually present for a given team, so any missing
# component's weight is automatically redistributed proportionally
# across the rest rather than just vanishing.
SACK_RATE_WEIGHT = 0.17
PRESSURE_RATE_WEIGHT = 0.34
ESPN_PBWR_WEIGHT = 0.34
PASS_SOS_WEIGHT = 0.15

RUN_BLOCK_COMPONENT_WEIGHT = 0.283  # stuff rate / YBC per att / ESPN RBWR, each
RUN_SOS_WEIGHT = 0.15

# Overall score weights -- see the module docstring. Small and deliberately
# so; penalty rate is a nudge, not a co-equal grading dimension.
OVERALL_PASS_WEIGHT = 0.46
OVERALL_RUN_WEIGHT = 0.46
OVERALL_PENALTY_WEIGHT = 0.08

# 13 EQUAL-WIDTH bands spanning the full 0-100 range (100/13 ~= 7.7 points
# each), not the traditional school scale where "passing" starts at 60 and
# F covers 0-59. That traditional scale assumes an absolute score (e.g. "%
# of questions right"); ours is a CURVED score built by averaging several
# min-max-normalized stats, which mathematically clusters most teams near
# 50 regardless of how good they actually are (averaging reduces spread).
# Mapped onto the traditional scale, that clustering would wrongly dump
# most of the league into "F". Equal-width bands centered on 50 = C instead
# give a spread of grades that actually looks like a report card.
GRADE_BANDS = [
    (92.3, "A+"), (84.6, "A"), (76.9, "A-"),
    (69.2, "B+"), (61.5, "B"), (53.8, "B-"),
    (46.2, "C+"), (38.5, "C"), (30.8, "C-"),
    (23.1, "D+"), (15.4, "D"), (7.7, "D-"),
    (0, "F"),
]


def score_to_letter(score: float | None) -> str | None:
    if score is None or pd.isna(score):
        return None
    for threshold, letter in GRADE_BANDS:
        if score >= threshold:
            return letter
    return "F"


def _normalize(series: pd.Series, higher_is_better: bool) -> pd.Series:
    """Min-max scale a column to 0-100 within whatever rows are passed in
    (the caller is responsible for passing in only one week's rows, i.e.
    one "league" to compare against)."""
    lo, hi = series.min(), series.max()
    if pd.isna(lo) or pd.isna(hi) or hi == lo:
        # Every team tied (or only one team present) -- nothing to compare,
        # so give everyone a neutral middle score rather than dividing by 0.
        return pd.Series(50.0, index=series.index)
    scaled = (series - lo) / (hi - lo) * 100
    return scaled if higher_is_better else 100 - scaled


def _weighted_average(components: list[tuple[pd.Series, float]]) -> pd.Series:
    """Weighted average of whichever components are non-null for each row.
    A missing component (e.g. ESPN data not entered yet) drops its weight
    out of the denominator entirely, so the remaining components'
    *relative* weights are preserved rather than the gap being silently
    averaged away unweighted. Passing equal weights (e.g. all 1s) gives a
    plain average, same as before."""
    weighted_sum = None
    weight_sum = None
    for series, weight in components:
        present = series.notna()
        contribution = series.fillna(0) * weight * present
        weighted_sum = contribution if weighted_sum is None else weighted_sum + contribution
        this_weight = present.astype(float) * weight
        weight_sum = this_weight if weight_sum is None else weight_sum + this_weight
    return weighted_sum / weight_sum.replace(0, pd.NA)


def compute_grades(
    team_week_stats: pd.DataFrame, espn_team_rates: pd.DataFrame, schedule_strength: pd.DataFrame
) -> pd.DataFrame:
    """
    team_week_stats: one row per team per week, for a single season, with
        columns team_abbr, week, dropbacks, sacks_allowed,
        pressure_rate_allowed, stuff_rate, yards_before_contact_per_att,
        penalty_rate.
    espn_team_rates: one row per team for that season (some teams may be
        missing if not entered yet), columns team_abbr,
        pass_block_win_rate, run_block_win_rate.
    schedule_strength: one row per team for that season (some teams may be
        missing early in a season), columns team_abbr, pass_sos_score,
        run_sos_score -- see pipeline/steps/pull_schedule_strength.py.

    Returns team_week_stats with the grade columns added.
    """
    df = team_week_stats.merge(espn_team_rates, on="team_abbr", how="left")
    df = df.merge(schedule_strength, on="team_abbr", how="left")

    sack_rate = df["sacks_allowed"] / df["dropbacks"].replace(0, pd.NA)

    results = []
    for week, week_df in df.groupby("week"):
        idx = week_df.index
        week_sack_rate = sack_rate.loc[idx]

        pass_components = [
            _normalize(week_sack_rate, higher_is_better=False),
            _normalize(week_df["pressure_rate_allowed"], higher_is_better=False),
            _normalize(week_df["pass_block_win_rate"], higher_is_better=True),
            _normalize(week_df["pass_sos_score"], higher_is_better=True),
        ]
        run_components = [
            _normalize(week_df["stuff_rate"], higher_is_better=False),
            _normalize(week_df["yards_before_contact_per_att"], higher_is_better=True),
            _normalize(week_df["run_block_win_rate"], higher_is_better=True),
            _normalize(week_df["run_sos_score"], higher_is_better=True),
        ]

        # Components built from an all-null column (e.g. nobody has entered
        # ESPN data yet, or schedule strength hasn't been computed yet)
        # score every team 50 by _normalize's tie-break rule, which would
        # silently pull every grade toward the middle. Mask those back out
        # to NaN so _weighted_average correctly ignores them.
        for col_name, components, component_idx in (
            ("pass_block_win_rate", pass_components, 2),
            ("pass_sos_score", pass_components, 3),
            ("run_block_win_rate", run_components, 2),
            ("run_sos_score", run_components, 3),
        ):
            if week_df[col_name].isna().all():
                components[component_idx] = pd.Series(pd.NA, index=idx)

        week_df = week_df.copy()
        pass_block_score = _weighted_average([
            (pass_components[0], SACK_RATE_WEIGHT),
            (pass_components[1], PRESSURE_RATE_WEIGHT),
            (pass_components[2], ESPN_PBWR_WEIGHT),
            (pass_components[3], PASS_SOS_WEIGHT),
        ])
        run_block_score = _weighted_average([
            (run_components[0], RUN_BLOCK_COMPONENT_WEIGHT),
            (run_components[1], RUN_BLOCK_COMPONENT_WEIGHT),
            (run_components[2], RUN_BLOCK_COMPONENT_WEIGHT),
            (run_components[3], RUN_SOS_WEIGHT),
        ])

        # Stretch each blended score to span the full 0-100 range across
        # this week's teams -- see the module docstring for why this step
        # exists (otherwise almost nobody ever hits a true A+ or F).
        week_df["pass_block_score"] = _normalize(pass_block_score, higher_is_better=True)
        week_df["run_block_score"] = _normalize(run_block_score, higher_is_better=True)

        penalty_score = _normalize(week_df["penalty_rate"], higher_is_better=False)
        if week_df["penalty_rate"].isna().all():
            penalty_score = pd.Series(pd.NA, index=idx)

        overall_raw = _weighted_average([
            (week_df["pass_block_score"], OVERALL_PASS_WEIGHT),
            (week_df["run_block_score"], OVERALL_RUN_WEIGHT),
            (penalty_score, OVERALL_PENALTY_WEIGHT),
        ])
        week_df["overall_score"] = _normalize(overall_raw, higher_is_better=True)
        results.append(week_df)

    out = pd.concat(results).sort_index()
    out["pass_block_grade"] = out["pass_block_score"].apply(score_to_letter)
    out["run_block_grade"] = out["run_block_score"].apply(score_to_letter)
    out["overall_grade"] = out["overall_score"].apply(score_to_letter)
    out["grade_formula_version"] = GRADE_FORMULA_VERSION

    return out
